stan/math/prim/prob/wiener5_lpdf.hpp

```diff
@@ -0,0 +1,428 @@
+
+#include <stan/math/prim/fun.hpp>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+enum class FunType { Density, GradT, GradA, GradV, GradW, GradSV, GradSW };
+
+/**
+ * Implementation function for calculating the density or gradient w.r.t. a
+ * specified parameter of the 5-parameter Wiener distribution
+ *
+ * @tparam FunTypeEnum An enum value indicating the quantity required
+ * @param y A scalar variable; the reaction time in seconds
+ * @param a The boundary separation
+ * @param vn The drift rate
+ * @param wn The relative starting point
+ * @param sv The inter-trial variability of the drift rate
+ * @param err The error tolerance for estimation
+ * @param use_log Whether to return calculation w.r.t. the log value
+ *
+ */
+template <FunType FunTypeEnum>
+inline double wiener5_helper(const double& y, const double& a, const double& vn,
+                             const double& wn, const double& sv,
+                             const double& err = log(1e-12),
+                             bool use_log = false) {
+  double w = 1.0 - wn;
+  double v = -vn;
+  double sv_sqr = square(sv);
+  double one_plus_svsqr_y = 1 + sv_sqr * y;
+  double two_avw = 2 * a * v * w;
+  double two_log_a = 2 * log(a);
+  double ans;
+  double ld = 0;
+
+  if (FunTypeEnum == FunType::GradSV || FunTypeEnum == FunType::GradV) {
+    ld = wiener5_helper<FunType::Density>(y, a, vn, wn, sv, err);
+    if (FunTypeEnum == FunType::GradV) {
+      ans = (a * (1 - wn) - vn * y);
+      if (sv != 0) {
+        ans /= 1 + sv_sqr * y;
+      }
+    } else {
+      double t1 = -y / one_plus_svsqr_y;
+      double t2 = (square(a * w) + two_avw * y + square(v * y))
+                  / square(one_plus_svsqr_y);
+      ans = sv * (t1 + t2);
+    }
+  } else {
+    double lg1;
+    if (sv != 0) {
+      lg1 = (sv_sqr * square(a * w) - two_avw - square(v) * y) / 2.0
+                / one_plus_svsqr_y
+            - two_log_a - 0.5 * log(one_plus_svsqr_y);
+    } else {
+      lg1 = (-two_avw - square(v) * y) / 2.0 - two_log_a;
+    }
+    double ans0 = 0;
+    if (FunTypeEnum != FunType::Density) {
+      double var_a = (FunTypeEnum == FunType::GradA) ? w : a;
+      double var_b = (FunTypeEnum == FunType::GradA) ? a : w;
+      if (sv != 0) {
+        if (FunTypeEnum == FunType::GradT) {
+          ans0 = -0.5
+                 * (square(sv_sqr) * (y + square(a * w))
+                    + sv_sqr * (1 - two_avw) + square(v))
+                 / square(one_plus_svsqr_y);
+        } else {
+          ans0 = (-v * var_a + sv_sqr * square(var_a) * var_b)
+                 / one_plus_svsqr_y;
+        }
+      } else {
+        ans0 = (FunTypeEnum == FunType::GradT) ? -0.5 * square(v) : -v * var_a;
+      }
+    }
+    double es = (err - lg1) + (FunTypeEnum == FunType::GradT) * two_log_a;
+    double two_es = 2.0 * es;
+    double y_asq = y / square(a);
+    double log_y_asq = log(y) - two_log_a;
+
+    double K1, u_eps, arg;
```

Comment:
We use multiple lines for instantiation

```suggestion
    double K1;
    double u_eps;
    double arg;
```

And you re-use K1 multiple times in the below giving it multiple definitions which is very confusing to keep track of. Why not just have multiple values? I think breaking this up into multiple functions for grad, lpdf, etc. will make this a lot clearer